src/psyclone/tests/dynamo0p3_transformations_test.py

```diff
@@ -6474,3 +6475,546 @@ def test_no_acc():
+    assert "CALL f2_proxy%halo_exchange_finish(depth=1)" in result
+    assert "IF (m1_proxy%is_dirty(depth=1)) THEN" in result
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+    #
+
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    #
+    rc_trans = Dynamo0p3RedundantComputationTrans()
+    loop = schedule.children[0]
+    rc_trans.apply(loop, depth=1)
+    result = str(psy.gen)
+    assert "CALL f1_proxy%halo_exchange_start(depth=1)" not in result
+    assert "CALL f1_proxy%halo_exchange_finish(depth=1)" not in result
+    assert "CALL f2_proxy%halo_exchange_start(depth=1)" in result
+    assert "CALL f2_proxy%halo_exchange_finish(depth=1)" in result
+    assert "IF (m1_proxy%is_dirty(depth=1)) THEN" in result
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+    #
+    loop = schedule.children[1]
+    rc_trans.apply(loop, depth=1)
+    result = str(psy.gen)
+    assert "CALL f1_proxy%halo_exchange_start(depth=1)" not in result
+    assert "CALL f1_proxy%halo_exchange_finish(depth=1)" not in result
+    assert "CALL f2_proxy%halo_exchange_start(depth=1)" not in result
+    assert "CALL f2_proxy%halo_exchange_finish(depth=1)" not in result
+    assert "IF (m1_proxy%is_dirty(depth=1)) THEN" in result
+    assert "CALL m1_proxy%halo_exchange(depth=1)" in result
+
+    if TEST_COMPILE:
+        # If compilation testing has been enabled (--compile flag
+        # to py.test)
+        assert code_compiles("dynamo0.3", psy, tmpdir, f90, f90flags)
+
+
+def test_rc_redund_async_halo_exchange(monkeypatch, tmpdir, f90, f90flags):
+    '''Test that an asynchronous halo exchange works correctly with
+    redundant computation being applied.
+    '''
+
+    # ensure we compute over annexed dofs so no halo exchanges are required
+    monkeypatch.setattr(_API_CONFIG, "_compute_annexed_dofs", True)
+    _, info = parse(os.path.join(
+        BASE_PATH, "14.7_halo_annexed.f90"),
+                    api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(info)
+    schedule = psy.invokes.invoke_list[0].schedule
+
+    # Make it so that halo exchanges are required to depth 2 for
+    # fields m1, m2, f1 and f2. m2 will have a set clean for depth 1
+    # after the last loop.
+    rc_trans = Dynamo0p3RedundantComputationTrans()
+    loop = schedule.children[2]
+    rc_trans.apply(loop, depth=2)
+
+    # make m2 halo exchange asynchronous and check depths and set
+    # clean are generated correctly for m2
+    ahex_trans = Dynamo0p3AsyncHaloExchangeTrans()
+    m2_hex = schedule.children[5]
+    schedule, _ = ahex_trans.apply(m2_hex)
+    result = str(psy.gen)
+    assert (
+        "      IF (m2_proxy%is_dirty(depth=2)) THEN\n"
+        "        CALL m2_proxy%halo_exchange_start(depth=2)\n"
+        "      END IF \n"
+        "      !\n"
+        "      IF (m2_proxy%is_dirty(depth=2)) THEN\n"
+        "        CALL m2_proxy%halo_exchange_finish(depth=2)\n"
```

Comment:
I guess this is the LFRic API rather than PSyclone but it seems strange that we have to pass in `depth` to `halo_exchange_finish`.

Reply:
Yes it is their API. I hadn't really thought about it, but could knowing the size of data being received e.g. to allow pre-allocation of appropriate buffer space could potentially improve performance?